### PR TITLE
[FEATURE] Nominatim locator filter and batch Nominatim geocoder algorithm

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR integrates Nominatim geocoding with QGIS through two features:

A new *Nominatim geocoder locator filter*, activated using the ‘nom’ prefix in the locator widget at the bottom-left corner of the QGIS main window. To insure QGIS users respect the usage policy of the  OpenStreetMap Foundation (OSMF)-maintained Nominatim endpoint, the filter does the following:
- In contrast to all of our current locator filter, the Nominatim filter delays its fetching of results of 1 second. This behavior is to avoid querying the API at every keystroke (which would be auto-completion search, explicitly prohibited by the policy).
- In addition to Qt’s network query caching mechanism, the Nominatim geocoder class has its own caching mechanism to minimize remote fetching.
- Upon triggering the first Nominatim result (by hitting enter), a one-time message bar will pop up and acknowledge the origin of the data and offer a link to Nominatim’s homepage.

Peek 2020-12-18 08-54
_Screencast showcasing the one-time data origin acknowledgement_

A new *Batch Nominatim geocoder algorithm* found in the Vector tools section of QGIS’ processing toolbox. Once again, to insure our users do not breach the usage policy, the algorithm does the following:
- Caps the number of queries/requests sent to the Nominatim to one per second (by forcing a 1,000ms wait in-between queries).
- As mentioned above, the Nominatim geocoder class has its own caching mechanism to minimize remote fetching. The cache will remember its content across multiple executions of the algorithm.
- A information message is printed in the algorithm’s log at every run reminding users of the ODbl license under which the data is provided.

image
_Information message appears in the log every time the algorithm is executed_

In addition, as required by the usage policy, all of our queries to the OSM-maintained Nominatim endpoint are identifiable via a User-Agent string ending with “ QGIS/{version_number}”. 

I believe that all of allows for QGIS to offer out-of-the-box geocoding to its users that is compliant with OSMF’s usage policy, as well as insuring a fair visibility of the data’s origin and license. This also prevents users from adopting alternatives (home-made scripts, QGIS plugins, etc.) that wouldn’t be as good as citizen as we are :wink:.

(For more details on the Nominatim geocoder class itself, see PR #40580)

### CODE REVIEW

**Comment by rduivenvoorde** [comment]
Nice @nirvn ! 

FYI: I did a Nominatim plugin (https://plugins.qgis.org/plugins/nominatim_locator_filter/) also a locator filter (using the 'osm' prefix :-) ) for Nominatim. If this is merged, I will remove that one.

My fix to not to fire too many requests was that the user needed to add a space at the end of the search string.

I think you did a very good work to honour the user-policies!

Maybe we should make the use/appeareance of the locator more prominent. Just yesterday I was asked (by a very experienced QGIS user) what my other locator plugin (to our national service) actually was doing... he was totally not aware of the locator functionality!

I'm eager to see the batch-part work, as I think (for other geocoders, mostly national ones) this is a really BIG thing in user land: being able to geocode a csv or excel with addresses to your national geocoder ...

Thanks for you work on this.

**Comment by nirvn** [comment]
Richard,

Thanks. I looked into the space/arbitrary character workaround, but thought
it was worth going the extra mile and add a delayed fetching mechanism to
the locator. With for e.g. space, you'd still end up sending multiple
queries as you are building your address (e.g. "street 20,my town,a
district" has at least three spaces). Having a 1 sec. delay (reset at every
stroke) is IMHO the best user experience as well as best mechanism to
maximize the chance of us sending the desired address as the query vs. part
of it.

The locator is indeed a gem :)

Nyall deserves most of the credit here, the geocodng foundation he has
added in upcoming 3.18 is great.

Mathieu

File: src/analysis/processing/qgsalgorithmbatchnominatimgeocode.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  : QgsBatchGeocodeAlgorithm( &mNominatimGeocoder )
+{
+}
+
+QString QgsBatchNominatimGeocodeAlgorithm::name() const
+{
+  return QStringLiteral( "batchnominatimgeocoder" );
+}
+
+QString QgsBatchNominatimGeocodeAlgorithm::displayName() const
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmbatchnominatimgeocode.cpp:37]: Can you also add the short help/descriptions too?

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 2]
```diff
@@ -1185,3 +1188,30 @@ void QgsGotoLocatorFilter::triggerResult( const QgsLocatorResult &result )
 
   mapCanvas->flashGeometries( QList< QgsGeometry >() << QgsGeometry::fromPointXY( point ) );
 }
+
+QgsNominatimLocatorFilter::QgsNominatimLocatorFilter( QgsGeocoderInterface *geocoder, QgsMapCanvas *canvas )
+  : QgsGeocoderLocatorFilter( QStringLiteral( "nominatimgeocoder" ), tr( "Nominatim Geocoder" ), QStringLiteral( "nom" ), geocoder, canvas )
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:1193]: If we're shipping this as the out-of-the-box place searching method, could we pick a nicer prefix? Maybe a generic single character option like ":" or ">" or ...?

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -3940,6 +3942,9 @@ void QgisApp::createStatusBar()
   mLocatorWidget->locator()->registerFilter( new QgsBookmarkLocatorFilter() );
   mLocatorWidget->locator()->registerFilter( new QgsSettingsLocatorFilter() );
   mLocatorWidget->locator()->registerFilter( new QgsGotoLocatorFilter() );
+
+  mNominatimGeocoder.reset( new QgsNominatimGeocoder() );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:3946]:
```suggestion
  mNominatimGeocoder = qgis::make_unique< QgsNominatimGeocoder> ();
```

**Comment by nyalldawson** [comment]
Looking good!

File: src/app/locator/qgsinbuiltlocatorfilters.cpp (hunk 2 above)
Comment (nirvn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:1193]: Hmm maybe a question mark?

Comment (nyalldawson) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:1193]: I thought about that, but we probably should keep ? free for a future "search for help on ...." filter...

Comment (nirvn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:1193]: OK, then I'd go for > , which can be seen as an arrow head, indicating we'll "go somewhere" :)

**Comment by nirvn** [comment]
@nyalldawson , review addressed. He's how the short description looks like:
image

**Comment by roya0045** [comment]
@nirvn instead of only accepting a field, could it also support expressions? If you have the address spread over 3 fields, need to add a static element, etc. This would be more versatile also in the modeler. Adding a specific is an extra step that results in the same thing but I think adding that flexibility from the onset may be good.

**Comment by pnorman** [comment]
A couple questions.

- The usage policy states "Apps must make sure that they can switch the service at our request at any time (in particular, switching should be possible without requiring a software update)." How is this done in QGIS?

- The usage limit is per application. How will you make sure across all qgis users that it does not exceed 1 req/s? What is the user experience if there are a lot of qgis users geocoding at once and someone has to wait?

- If you type something in does it automatically search even if you're partway through a word? That would be auto-completion even if it only does it every 1 second.


cc @lonvia

**Comment by rduivenvoorde** [comment]
> The usage limit is per application. How will you make sure across all qgis users that it does not exceed 1 req/s? What is the user experience if there are a lot of qgis users geocoding at once and someone has to wait?

Are you sure that is per application? Seems to me this is per User (per application of that user). I'm not aware of client applications which orchestrate there use over the globe to limit there calls to a server api.

**Comment by pnorman** [comment]
> Are you sure that is per application? Seems to me this is per User (per application of that user). I'm not aware of client applications which orchestrate there use over the globe to limit there calls to a server api.

Quoting from the policy, "the sum of traffic by all your users should not exceed the limits."

**Comment by nyalldawson** [comment]
So what's your conclusion @nirvn? Based on @pnorman 's feedback, is Nominatin completely incompatible with qgis? Or do you think there's any chance of a compromise here/special consideration for qgis?

**Comment by roya0045** [comment]
Having a limit of 1 request per second for all qgis instance seems like a logistical nightmare (would need to setup a universal queue on an independent machine to handle request), then what happens if the delays are long or someone make thousands of requests? Should request be interwoven, first come first serve and with possibly very long delays (or capping batch request to ensure reasonable times?). This will cause bad UX and may not be worth it if this feature becomes moderately used. :/

**Comment by nyalldawson** [comment]
@roya0045 I agree -- if there's no room to move in that requirement then that's a solid road block that qgis (and any qgis Nominatin plugin) can't possibly work around

**Comment by lonvia** [comment]
The restrictions are just for the OSMF instance of Nominatim. We only have so much hardware available and as much as we all love QGIS, you'd still be a rather large customer. There are other options, including running your own instance of Nominatim. We really should talk about this directly. @mbernasocchi already got in touch and I hope we can chat about this some time next year.

**Comment by mbernasocchi** [comment]
@ionvia thanks a lot will answer your mail in the next days! Happy New year

**Comment by nirvn** [comment]
Still hoping it's not DOA.

**Comment by roya0045** [comment]
@nirvn TBH there is already at least one plugim using it, if that is not currently an issue maybe your work could be fine.

**Comment by elpaso** [comment]
This has been discussed by the PSC and we are trying to find a solution. 

As far as the plugin is concerned (if you are referring to my ancient geocoding plugin) it was developed on purpose to not allow batch geocoding or live-search, even if you are very fast to type and click you cannot make more than one request per second.

**Comment by roya0045** [comment]
@elpaso No I'm pretty sure there is at least another one.

**Comment by rduivenvoorde** [comment]
I think Alex means my Nominatim Locator Filter: https://plugins.qgis.org/plugins/nominatim_locator_filter/
I tried to not use 'suggest' by asking the user to append a space after their search string.

**Comment by rduivenvoorde** [comment]
I had a short look on how to install Nominatim on one of our servers, but... it's not that easy... Most notably because you need to load the whole world dataset into a db etc... But maybe with some help of OSM-admins and some information from the about hardware-requirements, we can think about hosting one instance ourselves? Or sponsor one dedicated node in the OSM-hardware farms?

**Comment by jef-n** [comment]
> The restrictions are just for the OSMF instance of Nominatim. We only have so much hardware available and as much as we all love QGIS, you'd still be a rather large customer. There are other options, including running your own instance of Nominatim. We 
really should talk about this directly. @mbernasocchi already got in touch and I hope we can chat about this some time next year.

We wouldn't be a customer at all, would we?  "We" are not running QGIS (except for tests maybe) the users do.   But there might be larger orgs running QGIS too, but most will probably come from individuals - so just small "customers", but potentially many.

The policy is a bit vague on what apps and websites are.  Ie. if multiple independent websites running the same software, count as one or individually?  If there was library/framework to access nominatim (not sure if there is), would all users of it also count in sum even if the framework is used in different applications/on different sites? For the website case, what if the access directly happens from the client-side and not though the website?   And how would we make sure that any restriction, we would implement into the QGIS to support the policy, is not removed by users?

Just thinking out loud - of course you are free to offer the service as you like.   But I'm still wondering if it isn't better to support the existing infrastructure to support more clients, than to built our own.   IMHO we should focus on the software.

**Comment by nirvn** [comment]
Having QGIS run/manage our own instance feels like non starter to me. I'd rather maintain our energy and focus on the task of building a great open source GIS software. I'd also find it far more logical for us to find ways for OSMF to support their already up and running nominatim instead. We'd continue doing what we do best in both of our respective project.

If OSMF is apprehensive of the possible strain on the  well-behaved bulk geocoder shipped by default, we could merge only the locator feature for 3.18 and keep the discussion running for the bulk geocoder algorithm. ATM, I go to my browser to do nominatim searches, which involves loading osm.org each time and therefore ending up costing more CPU cycles on the OSM servers than merely making a nominatim API call within QGIS.

As for the bulk geocoder, I still see the positive of having a good citizen easily accessible to users instead of them scripting solutions via 3rd party python modules et cie. None of which have a mechanism to cap requests across all uses of given python modules either :)

**Comment by nyalldawson** [comment]
Marking as freeze exempt for a week given the original submission date of this PR

**Comment by nirvn** [comment]
Hope is fading away. I'll close this by the time we release QGIS 3.18.

**Comment by gioman** [comment]
> Hope is fading away.

Really sad as this is no brainier win win for both osm and qgis.

**Comment by nirvn** [comment]
Alright, I've lost all hopes, closing. What an unfortunate outcome this is.

**Comment by rduivenvoorde** [comment]
Just a thought...

@nirvn I see more and more 'national locators' appear: python implementation of the 'GeocoderFilter'  interface (like we have for The Netherlands and I created a locator plugin for: https://github.com/rduivenvoorde/pdoklocatorplugin).

Would it maybe possible to 'merge/morph' your ideas in this work with this kind of 'GeocoderFilter' implementations?

Though this probably is not easy, because it would mean that c++ code would be calling these Python implementations?

**Comment by gioman** [comment]
> Alright, I've lost all hopes, closing. What an unfortunate outcome this is.

@nirvn I think this should be reopened and discussion continue.

**Pull request closed without merging**